### PR TITLE
Adding ova deployment feature

### PR DESCRIPTION
**Pull request by sumitAgrawal007**
### Description

Adding feature to deploy a VM from OVA template

### Acceptance tests
- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch? (If so, please include the test log in a gist)

https://gist.github.com/sumitAgrawal007/42e1f54fe6ed9a91466b1b14aa81a554

### CODE REVIEW

File: vsphere/internal/helper/ovfdeploy/ovf_helper.go [hunk 1]
```diff
@@ -38,8 +41,8 @@ func (pr *ProgressReader) Read(p []byte) (n int, err error) {
 	return
 }
 
-func DeployOVFAndGetResult(ovfCreateImportSpecResult *types.OvfCreateImportSpecResult, resourcePoolObj *object.ResourcePool,
-	folder *object.Folder, host *object.HostSystem, OvfPath string, ovfFromLocal bool) error {
+func DeployOvfAndGetResult(ovfCreateImportSpecResult *types.OvfCreateImportSpecResult, resourcePoolObj *object.ResourcePool,
+	folder *object.Folder, host *object.HostSystem, filePath string, isDeployOva bool, fromLocal bool, allowUnverifiedSSL bool) error {
```
Comment (shylasrinivas) [inline vsphere/internal/helper/ovfdeploy/ovf_helper.go:45]: rename isDeployOva to deployOva

File: vsphere/internal/helper/ovfdeploy/ovf_helper.go [hunk 2]
```diff
@@ -197,6 +180,178 @@ func Upload(ctx context.Context, item types.OvfFileItem, f io.Reader, url string
...
+				return "", err
+			}
+			defer resp.Body.Close()
+
+			if resp.StatusCode == http.StatusOK {
+				bodyBytes, err := ioutil.ReadAll(resp.Body)
+				if err != nil {
+					return "", err
+				}
+				ovfDescriptor = string(bodyBytes)
```
Comment (shylasrinivas) [inline vsphere/internal/helper/ovfdeploy/ovf_helper.go:278]: indentation seems off in all new functions, can you check?

File: vsphere/internal/vmworkflow/virtual_machine_ovfdeploy_subresource.go [hunk 3]
```diff
@@ -44,5 +43,11 @@ func VirtualMachineOVFDeploySchema() map[string]*schema.Schema {
 			Elem:        &schema.Schema{Type: schema.TypeString},
 			ForceNew:    true,
 		},
+		"allow_unverified_ssl_cert": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     true,
+			Description: "If to allow unverified ssl certificates while deploying ovf/ova from url.",
```
Comment (shylasrinivas) [inline vsphere/internal/vmworkflow/virtual_machine_ovfdeploy_subresource.go:50]: Change verbiage - remove "if to"

File: vsphere/resource_vsphere_virtual_machine.go [hunk 4]
```diff
@@ -949,11 +953,14 @@ func resourceVSphereVirtualMachineCustomizeDiff(d *schema.ResourceDiff, meta int
 		remoteOvfUrl := d.Get("ovf_deploy.0.remote_ovf_url").(string)
 
 		if localOvfPath == "" && remoteOvfUrl == "" {
-			return fmt.Errorf("either local ovf path or remote ovf url is required, both can't be empty")
+			return fmt.Errorf("either local ovf/ova path or remote ovf/ova url is required, both can't be empty")
+		}
+		if localOvfPath != "" && remoteOvfUrl != "" {
+			return fmt.Errorf("both local ovf/ova path and remote ovf/ova url is given, please specify only one source")
```
Comment (shylasrinivas) [inline vsphere/resource_vsphere_virtual_machine.go:959]: is given --> are provided

File: vsphere/resource_vsphere_virtual_machine.go [hunk 5]
```diff
@@ -1375,30 +1392,13 @@ func resourceVsphereMachineDeployOVF(d *schema.ResourceData, meta interface{}) (
...
-		if resp.StatusCode == http.StatusOK {
-			bodyBytes, err := ioutil.ReadAll(resp.Body)
-			if err != nil {
-				return nil, err
-			}
-			ovfDescriptor = string(bodyBytes)
-		}
+	ovfDescriptor, err := ovfdeploy.GetOvfDescriptor(filePath, isDeployOva, fromLocal, allowUnverifiedSSL)
+	if err != nil {
+		return nil, fmt.Errorf("error while reading the ovf File %s, %s ", filePath, err)
```
Comment (shylasrinivas) [inline vsphere/resource_vsphere_virtual_machine.go:1397]: typo - File -> file

File: website/docs/r/virtual_machine.html.markdown [hunk 6]
```diff
@@ -245,18 +245,21 @@ resource "vsphere_virtual_machine" "vm" {
...
-### Deploying VM from an OVF template
-OVF templates can be deployed both from local system and remote URL into the 
+### Deploying VM from an OVF/OVA template
+Ovf and ova templates can be deployed both from local system and remote URL into the 
 vcenter using the `ovf_deploy` property. When deploying from local system, the 
-path to the .ovf template needs to be given and all other necessary files like .vmdk 
-files also should be present in the same directory as the .ovf file. While deploying, 
-the VM properties like `name`, `datacenter_id`, `resource_pool_id`, `datastore_id`, 
-`host_system_id`, `folder`, `vapp` can only be set. All other VM properties are taken from the OVF 
+path to the ovf or ova template needs to be given. While deploying ovf, all other
```
Comment (shylasrinivas) [inline website/docs/r/virtual_machine.html.markdown:251]: given --> provided

File: website/docs/r/virtual_machine.html.markdown [hunk 7]
```diff
@@ -1328,34 +1344,33 @@ resource "vsphere_virtual_machine" "vm" {
...
 ~> **NOTE:** Either `local_ovf_path` or `remote_ovf_url` is required, both can't be empty.
 
 * `ip_allocation_policy` - (Optional) The IP allocation policy.
 * `ip_protocol` - (Optional) The IP protocol.
 * `disk_provisioning` - (Optional) The disk provisioning. If set, all the disks in the deployed OVF will have 
-   the same specified disk type (e.g., thin provisioned).
+   the same specified disk type (accepted values {thin, flat, thick, sameAsSource}).
 * `ovf_network_map` - (Optional) The mapping of name of network identifiers from the ovf descriptor to network UUID in the 
    VI infrastructure.
+* `allow_unverified_ssl_cert` - (Optional) If to allow unverified ssl certificates while deploying ovf/ova from url.
```
Comment (shylasrinivas) [inline website/docs/r/virtual_machine.html.markdown:1372]: Remove "If to"

**Comment by sumitAgrawal007** [comment]
@bill-rich Hi, Can you review this.

**Pull request merged**